store/tikv/backoff.go

```diff
@@ -59,12 +57,88 @@ func NewBackoff(retry, base, cap, jitter int) func() error {
 		time.Sleep(time.Duration(sleep) * time.Millisecond)
 
 		attempts++
-		totalSleep += sleep
 		lastSleep = sleep
-		return nil
+		return lastSleep
 	}
 }
 
 func expo(base, cap, n int) int {
 	return int(math.Min(float64(cap), float64(base)*math.Pow(2.0, float64(n))))
 }
+
+type backoffType int
+
+const (
+	boTiKVRPC backoffType = iota
+	boTxnLock
+	boPDRPC
+	boRegionMiss
+)
+
+func (t backoffType) createFn() func() int {
+	switch t {
+	case boTiKVRPC:
+		return NewBackoffFn(100, 2000, EqualJitter)
+	case boTxnLock:
+		return NewBackoffFn(300, 3000, EqualJitter)
+	case boPDRPC:
+		return NewBackoffFn(500, 3000, EqualJitter)
+	case boRegionMiss:
+		return NewBackoffFn(100, 500, NoJitter)
+	}
+	return nil
+}
+
+const (
+	copBuildTaskMaxBackoff = 3000
+	newTxnMaxBackoff       = 3000
+	storeVersionMaxBackoff = 3000
+	scannerNextMaxBackoff  = 3000
+	batchGetMaxBackoff     = 10000
+	copNextMaxBackoff      = 10000
+	getMaxBackoff          = 10000
+	commitMaxBackoff       = 15000
+)
+
+// Backoff is a utility for retrying queries.
+type Backoff struct {
+	mu         sync.Mutex
+	fn         map[backoffType]func() int
+	maxSleep   int
+	totalSleep int
+	errors     []string
+}
+
+// NewBackoff creates a Backoff with maximum sleep time(in ms).
+func NewBackoff(maxSleep int) *Backoff {
+	return &Backoff{
+		maxSleep: maxSleep,
+	}
+}
+
+// Backoff sleeps a while base on the backoffType and records the error message.
+// It returns an retryable error if total sleep time exceeds maxSleep.
+func (b *Backoff) Backoff(typ backoffType, err string) error {
```

Comment:
The function name is the same as the type name, which is confusing.
How about rename the type to `Backoffer`?